src/repo/kv.rs

```diff
@@ -101,234 +95,301 @@ impl DataStore for KvDataStore {
 
-        batch.remove(recursive_key.as_str());
+            let recursive_key = get_pin_key(target, &PinMode::Recursive);
+            tx_tree.remove(recursive_key.as_str())?;
 
-        for cid in &set {
-            let already_pinned = get_pinned_mode(self, cid)?;
+            for cid in &set {
+                let already_pinned = get_pinned_mode(tx_tree, cid)?;
 
-            match already_pinned {
-                Some(PinMode::Recursive) | Some(PinMode::Direct) => continue, // this should be unreachable
-                Some(PinMode::Indirect) => {
-                    let indirect_key = get_pin_key(cid, &PinMode::Indirect);
-                    batch.remove(indirect_key.as_str());
+                match already_pinned {
+                    Some(PinMode::Recursive) | Some(PinMode::Direct) => continue, // this should be unreachable
+                    Some(PinMode::Indirect) => {
+                        // FIXME: not really sure of this but it might be that recursive removed
+                        // the others...?
+                        let indirect_key = get_pin_key(cid, &PinMode::Indirect);
+                        tx_tree.remove(indirect_key.as_str())?;
+                    }
+                    None => {}
                 }
-                None => {}
             }
-        }
 
-        Ok(self._apply_batch(batch)?)
+            Ok(())
+        });
+
+        launder(res)?;
+
+        self.flush_async().await
     }
 
     async fn list(
         &self,
-        expected_mode: Option<PinMode>,
+        requirement: Option<PinMode>,
     ) -> futures::stream::BoxStream<'static, Result<(Cid, PinMode), Error>> {
         let db = self.get_db();
 
-        // the minimum cid of version 0
-        let min_key = "pin.0.0000000000000000000000000000000000000000000000";
-        assert_eq!(min_key.len(), 52);
-
-        let iter = db.range(min_key..);
-        let mut all_keys: Vec<String> = vec![];
-
-        for item in iter.filter(|item| item.is_ok()) {
-            let (raw_key, _) = item.unwrap();
-            let key = String::from_utf8_lossy(raw_key.as_ref());
-
-            if !key.starts_with("pin.") {
-                continue;
-            }
-
-            all_keys.push(key.to_owned().to_string());
-        }
-
-        let st = async_stream::try_stream! {
-            for key in all_keys.iter() {
-                let cid_str_with_prefix = &key[4..];
-                let cid_str = &key[6..];
-
-                let pin_mode = match cid_str_with_prefix {
-                    _ if cid_str_with_prefix.starts_with("d") => {
-                        PinMode::Direct
-                    },
-
-                    _ if cid_str_with_prefix.starts_with("r") => {
-                        PinMode::Recursive
-                    }
-
-                    _ if cid_str_with_prefix.starts_with("i") => {
-                        PinMode::Indirect
+        // FIXME: this is still blocking ... might not be a way without a channel
```

Comment:
> mainly used for GC, there will not be many concurrent call to it

Since it's available from the http api, I'd consider it beign callable at all times. For GC... I think the transactions might be helping out with it a lot. However, I am not too sure if we should attempt to build a similar GC than what go-ipfs has (not sure what is the present implementation) but I was thinking of trying to go through more of the route of maintaining a collectable set at all times through reference counting.

The reference counting would of probably require serializing all of the documents as graphs inside the blockstore, paving a way for a more monolithic store instead of trying to go with a number of different stores which are pluggable. I can't really see an design of such composed store which would be anywhere near consistent or durable.

> so maybe a new thread can be spawned by something like `threadpool::spawn`

tokio already has [`spawn_blocking`](https://docs.rs/tokio/1.0.1/tokio/task/fn.spawn_blocking.html) so I'd rather not entangle a new threadpool implementation here. I did not yet use the `spawn_blocking` as that'd mess up the indentation at least and make this PR even less reviewable. I'll do this in a follow-up.

> I wonder why the PinStore and DataStore should be coupled

There is no particular reason, this was more of a "convenient" quick decision I made to enable hacking the memory and fs based stores together, getting the "ball rolling" so to speak. However, I am still thinking that perhaps we should just put everything of `Repo` behind a single monolithic sled/database implementation and not even try to view these as separate. Perhaps things like non-local storage of blocks (for example in a s3) should be viewed as extensions to a monolithic database backend rather than "interchangeable blocks". Abstracting over databases is really difficult either way.

> pin data be separatedly put in a standalone sled tree. Then the filter of key that is should start with "pin." can be removed.

My thoughts exactly! Tried to keep a lid on the number of changes which already spiraled out of control so kept the current way :) I tried to write some more ideas in TODOs of the `get_pin_key(cid: &Cid, mode: &Mode) -> String` fn.